mm/mm_gran/mm_granalloc.c

```diff
@@ -25,226 +25,46 @@
-                   * in the upper 16 bits.
-                   */
-
-                  shift = 16;
-                }
-
-              /* We know that the first free bit is now within the lower 16
-               * bits of 'curr'.  Is it in the upper or lower byte?
-               */
-
-              else if ((curr & 0x0000ff) == 0x000000ff)
-                {
-                  /* Not in the lower 8 bits.  The first free bit must be in
-                   * the upper 8 bits.
-                   */
-
-                  shift = 8;
-                }
-
-              /* We know that the first free bit is now within the lower 4
-               * bits of 'curr'.  Is it in the upper or lower nibble?
-               */
-
-              else if ((curr & 0x00000f) == 0x0000000f)
-                {
-                  /* Not in the lower 4 bits.  The first free bit must be in
-                   * the upper 4 bits.
-                   */
-
-                  shift = 4;
-                }
-
-              /* We know that the first free bit is now within the lower 4
-               * bits of 'curr'.  Is it in the upper or lower pair?
-               */
-
-              else if ((curr & 0x000003) == 0x00000003)
-                {
-                  /* Not in the lower 2 bits.  The first free bit must be in
-                   * the upper 2 bits.
-                   */
-
-                  shift = 2;
-                }
-
-              /* We know that the first free bit is now within the lower 4
-               * bits of 'curr'.  Check if we have the allocation at this
-               * bit position.
-               */
-
-              else if ((curr & mask) == 0)
-                {
-                  /* Yes.. mark these granules allocated */
-
-                  gran_mark_allocated(priv, alloc, ngranules);
-
-                  /* And return the allocation address */
-
-                  gran_leave_critical(priv);
-                  return (FAR void *)alloc;
-                }
-
-              /* The free allocation does not start at this position */
-
-              else
-                {
-                  shift = 1;
-                }
-
-              /* Set up for the next time through the loop.  Perform a 64
-               * bit shift to move to the next gran position and increment
-               * to the next candidate allocation address.
-               */
-
-              alloc  += (shift << priv->log2gran);
-              curr    = (curr >> shift) | (next << (32 - shift));
-              next  >>= shift;
-              bitidx += shift;
-            }
-        }
-
-      gran_leave_critical(priv);
-    }
-
-  return NULL;
+  FAR gran_t *gran = (FAR gran_t *)handle;
+  size_t ngran;
+  int posi;
+  int ret;
+  uintptr_t retp;
+
+  DEBUGASSERT(gran);
+  ngran = NGRANULE(gran, size);
+  if (!ngran || ngran > gran->ngranules) return NULL;
+
+  ret = gran_enter_critical(gran);
+  if (ret < 0)    return NULL;
```

Comment:
Did you run checkpatch? This kind of coding style is forbidden, it needs use use bracket as in MISRA-C

Reply:
Yes I did. but I will add brackets though I dislike long files.

Reply:
Far enough! :-D

Single lines with brackets are safe!